### DISCUSSION TITLE
Request search and replace functions

### DISCUSSION THREAD

**Discussion by swordream** [Milkdown/milkdown#412] [Ideas]
Implement search,replace,replaceall and so on,i think it is a commonly used functions。

能不能实现搜索和替换插件，感觉这是个非常必要且高频使用的功能。代码中可以通过

···
view.dispatch(
        state.tr.replace(
          tr.selection.from,
          tr.selection.from,
          new Slice(doc.content, 0, 0)
        )
···

这样来实现，我的意思是应该实现一个插件，默认提供搜索替换界面和功能。

**Comment by Saul-Mirone** [comment]
I don't have plan to implement it officially. You can create a community plugin.

**Comment by sytolk** [comment]
I have implement search and this works in Milkdown editorViewOptionsCtx editable mode = true. But how to setSelection on searchText when  editable mode = false @Saul-Mirone ?
```javascript
function searchFromSelection(
    view: EditorView,
    searchText: string,
    from: number,
    to: number
  ): boolean {
    const { state, dispatch } = view;
    const { doc } = state;
    let tr = state.tr;
    let found = false;

    doc.descendants((node, pos) => {
      if (found || !node.isText) return;

      if (pos >= from) {
        const text = caseSensitive ? node.text! : node.text!.toLowerCase();
        const index = text.indexOf(searchText, pos === from ? to - pos : 0);

        if (index !== -1) {
          const start = pos + index;
          const end = start + searchText.length;

          // Select the found text
          tr = tr.setSelection(TextSelection.create(doc, start, end));
          tr = tr.scrollIntoView();
          view.dom.focus();

          found = true;
        }
      }
    });

    if (found) {
      dispatch(tr);
    }

    return found;
  }
```

**Reply by sytolk** [reply]
I have use HTML selection API for this: https://github.com/tagspaces/tagspaces-common/blob/develop/packages/tagspaces-md/src/components/dialogs/HtmlSearcher.ts

feels free to use the code and create Search and Search/Replace milkdown plugin

**Reply by Saul-Mirone** [reply]
Prosemirror doesn't allow `setSelection` in readonly mode so I guess your solution is the only possible solution.
